Mathlib/Data/Matroid/Circuit.lean

```diff
@@ -0,0 +1,173 @@
+  fun h' ↦ h'.ne (hC.eq_of_dep_subset hC'.dep h'.subset)
+
+lemma Circuit.eq_of_subset_circuit (hC : M.Circuit C) (hC' : M.Circuit C') (h : C ⊆ C') : C = C' :=
+  hC'.eq_of_dep_subset hC.dep h
+
+lemma circuit_iff_dep_forall_diff_singleton_indep :
+    M.Circuit C ↔ M.Dep C ∧ ∀ e ∈ C, M.Indep (C \ {e}) := by
+  rw [circuit_iff_forall_ssubset, and_congr_right_iff]
+  refine fun _ ↦ ⟨fun h e heC ↦ h (Iff.mpr diff_singleton_sSubset heC), fun h I hIC ↦ ?_⟩
+  obtain ⟨e, he⟩ := exists_of_ssubset hIC
+  exact (h e he.1).subset (subset_diff_singleton hIC.subset he.2)
+
+/-! ### Independence and bases -/
+
+lemma Indep.insert_circuit_of_forall (hI : M.Indep I) (heI : e ∉ I) (he : e ∈ M.closure I)
+    (h : ∀ f ∈ I, e ∉ M.closure (I \ {f})) : M.Circuit (insert e I) := by
+  rw [circuit_iff_dep_forall_diff_singleton_indep, hI.insert_dep_iff, and_iff_right ⟨he, heI⟩]
+  rintro f (rfl | hfI)
+  · simpa [heI]
+  rw [← insert_diff_singleton_comm (by rintro rfl; contradiction),
+    (hI.diff _).insert_indep_iff_of_not_mem (by simp [heI])]
+  exact ⟨mem_ground_of_mem_closure he, h f hfI⟩
+
+lemma Indep.insert_circuit_of_forall_of_nontrivial (hI : M.Indep I) (hInt : I.Nontrivial)
+    (he : e ∈ M.closure I) (h : ∀ f ∈ I, e ∉ M.closure (I \ {f})) : M.Circuit (insert e I) := by
+  refine hI.insert_circuit_of_forall (fun heI ↦ ?_) he h
+  obtain ⟨f, hf, hne⟩ := hInt.exists_ne e
+  exact h f hf (mem_closure_of_mem' _ (by simp [heI, hne.symm]))
+
+lemma Circuit.diff_singleton_basis (hC : M.Circuit C) (he : e ∈ C) : M.Basis (C \ {e}) C := by
+  nth_rw 2 [← insert_eq_of_mem he]
+  rw [← insert_diff_singleton, (hC.diff_singleton_indep he).basis_insert_iff,
+    insert_diff_singleton, insert_eq_of_mem he]
+  exact Or.inl hC.dep
+
+lemma Circuit.basis_iff_eq_diff_singleton (hC : M.Circuit C) :
+    M.Basis I C ↔ ∃ e ∈ C, I = C \ {e} := by
+  refine ⟨fun h ↦ ?_, ?_⟩
+  · obtain ⟨e, he⟩ := exists_of_ssubset
+      (h.subset.ssubset_of_ne (by rintro rfl; exact hC.dep.not_indep h.indep))
+    exact ⟨e, he.1, h.eq_of_subset_indep (hC.diff_singleton_indep he.1)
+      (subset_diff_singleton h.subset he.2) diff_subset⟩
+  rintro ⟨e, he, rfl⟩
+  exact hC.diff_singleton_basis he
+
+lemma Circuit.basis_iff_insert_eq (hC : M.Circuit C) :
+    M.Basis I C ↔ ∃ e ∈ C \ I, C = insert e I := by
+  rw [hC.basis_iff_eq_diff_singleton]
+  refine ⟨fun ⟨e, he, hI⟩ ↦ ⟨e, ⟨he, fun heI ↦ (hI.subset heI).2 rfl⟩, ?_⟩,
+    fun ⟨e, he, hC⟩ ↦ ⟨e, he.1, ?_⟩⟩
+  · rw [hI, insert_diff_singleton, insert_eq_of_mem he]
+  rw [hC, insert_diff_self_of_not_mem he.2]
+
+/-! ### Closure -/
+
+lemma Circuit.closure_diff_singleton_eq (hC : M.Circuit C) (e : α) :
+    M.closure (C \ {e}) = M.closure C :=
+  (em (e ∈ C)).elim
+    (fun he ↦ by rw [(hC.diff_singleton_basis he).closure_eq_closure])
+    (fun he ↦ by rw [diff_singleton_eq_self he])
+
+lemma Circuit.subset_closure_diff_singleton (hC : M.Circuit C) (e : α) :
+    C ⊆ M.closure (C \ {e}) := by
+  by_cases he : e ∈ C
```

Comment:
It seems like either this or the previous should be easy to prove in terms of the other?